### PR TITLE
feat: Cross Program Line Listing API [DHIS2-13779] [2.40]

### PR DESCRIPTION
**Pull request by gnespolino**
Let's just review that this branch reflects all our work up to date on CPL linelisting app

### CODE REVIEW

**Review by d-bernat — APPROVED** [review]

**Review by d-bernat — APPROVED** [review]

**Review by d-bernat — APPROVED** [review]

**Comment by larshelge** [comment]
Closing due to inactivity.

**Comment by gnespolino** [comment]
@larshelge we're still working on this, so we need this still open.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/processing/CommonQueryRequestMapper.java [hunk 1]
```diff
@@ -97,6 +103,22 @@ public CommonParams map( CommonQueryRequest request )
...
+
+        programs.stream()
+            .flatMap( program -> getProgramAttributes( List.of( program ) )
+                .map( BaseIdentifiableObject::getUid )
+                // we need fully qualified dimension identifiers
+                .map( attributeUid -> Pair.of( program, attributeUid ) ) )
+            .forEach( fqDimension -> dimensionsByUid.put(
+                fqDimension.getRight(), fqDimension.getLeft().getUid() + "." + fqDimension.getRight() ) );
+
+        // removes all items already existing for which exists a FQ dimension name
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/processing/CommonQueryRequestMapper.java:117]: Suggestion:
Removes all existing items which already have a fully qualified dimension name.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/processing/CommonQueryRequestMapper.java [hunk 2]
```diff
@@ -97,6 +103,22 @@ public CommonParams map( CommonQueryRequest request )
...
+        request.getDimension().removeIf( dimensionsByUid::containsKey );
+        // adds all dimensions from programs
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/processing/CommonQueryRequestMapper.java:119]: Suggestion:
Adds all dimensions from all programs.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/processing/CommonQueryRequestMapper.java [hunk 3]
```diff
@@ -97,6 +103,22 @@ public CommonParams map( CommonQueryRequest request )
...
 
+        Map<String, String> dimensionsByUid = request.getDimension().stream()
+            .collect( Collectors.toMap( identity(), identity() ) );
+
+        programs.stream()
+            .flatMap( program -> getProgramAttributes( List.of( program ) )
+                .map( BaseIdentifiableObject::getUid )
+                // we need fully qualified dimension identifiers
+                .map( attributeUid -> Pair.of( program, attributeUid ) ) )
+            .forEach( fqDimension -> dimensionsByUid.put(
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/processing/CommonQueryRequestMapper.java:114]: Suggestion:
Maybe rename "fqDimension" to "fullDimension" or "fullyQualifiedDimension".
Initially, I could not understand what "fq" stood for.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/StringDimensionIdentifier.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+
+import org.apache.commons.lang3.tuple.Triple;
+
+@RequiredArgsConstructor( access = AccessLevel.PRIVATE )
+public class StringDimensionIdentifier
+{
+
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/StringDimensionIdentifier.java:38]: This empty line can be removed.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/StringDimensionIdentifier.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ */
+package org.hisp.dhis.analytics.common.dimension;
+
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+
+import org.apache.commons.lang3.tuple.Triple;
+
+@RequiredArgsConstructor( access = AccessLevel.PRIVATE )
+public class StringDimensionIdentifier
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/StringDimensionIdentifier.java:36]: Nice to have Javadoc for this class.

**Review by maikelarabori — APPROVED** [review]

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tei/TeiQueryController.java [hunk 6]
```diff
@@ -104,7 +109,9 @@
     @Nonnull
     private final ContextUtils contextUtils;
 
-    @GetMapping( "query/{trackedEntityType}" )
+    @GetMapping( value = "query/{trackedEntityType}", produces = { APPLICATION_JSON_VALUE,
+        "application/javascript" } )
+    @ResponseBody
```
Comment (gnespolino) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tei/TeiQueryController.java:114]: When the class is annotated with @RestController, then @ResponseBody is implicit

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tei/TeiQueryController.java [hunk 7]
```diff
@@ -127,6 +134,32 @@ Grid getGrid(
...
+            .build();
+
+        commonQueryRequestValidator.validate( queryRequest.getCommonQueryRequest() );
+        teiQueryRequestValidator.validate( queryRequest );
+
+        contextUtils.configureResponse( response, CONTENT_TYPE_JSON, RESPECT_SYSTEM_SETTING );
+
+        TeiQueryParams params = mapper.map( queryRequest );
+
+        return teiAnalyticsQueryService.getGridExplain( params );
```
Comment (gnespolino) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tei/TeiQueryController.java:160]: we might extract this method in a common private one, and invoke from each controller's methods

**Pull request merged**